### ISSUE TITLE
ThenBy after select (with previous orderby) throw exception

### ISSUE THREAD

**Issue by puschie286** [dotnet/efcore#34363]
## File a bug

Using ThenBy on ordered query where one entity was selected throws an exception.

If only one OrderBy is used (inner or outer), it works fine. If both OrderBy and ThenBy are done before Select, it works too.

https://dotnetfiddle.net/hiEFxJ

The goal is to (dynamically) join on multiple entities and order by properties of these entities. The join target only knows the source/base entity

Query code from example:
```C#
// part 1
var query = context.Set<Blog>()
	.Join( context.Set<Post>(), outer => new { Id = outer.BlogId }, inner => new { Id = inner.BlogId }, ( outer, inner ) => new { o = outer, i = inner } )
	.OrderBy( x => x.i.Title )
	.Select( x => x.o );
// part 2		
if( query is IOrderedQueryable<Blog> orderedQuery )
{
	query = orderedQuery.ThenBy( x => x.Url );
}
```

### Include stack traces

```
Unhandled exception. System.ArgumentException: Expression of type 'System.Linq.IQueryable`1[Blog]' cannot be used for parameter of type 'System.Linq.IOrderedQueryable`1[Blog]' of method 'System.Linq.IOrderedQueryable`1[Blog] ThenBy[Blog,String](System.Linq.IOrderedQueryable`1[Blog], System.Linq.Expressions.Expression`1[System.Func`2[Blog,System.String]])' (Parameter 'arg0')
   at System.Dynamic.Utils.ExpressionUtils.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arguments, ParameterInfo pi, String methodParamName, String argumentParamName, Int32 index)
   at System.Linq.Expressions.Expression.Call(Expression instance, MethodInfo method, Expression arg0, Expression arg1)
   at System.Linq.Queryable.ThenBy[TSource,TKey](IOrderedQueryable`1 source, Expression`1 keySelector)
   at Program.Main()
```

### Include provider and version information

(might be different for the fiddle)
EF Core version: ef 8.0.4
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: 8.0
Operating system: win10
IDE: rider

Expected SQL:
```sql
SELECT [b].[BlogId], [b].[Url]
FROM [Blogs] AS [b]
INNER JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
ORDER BY [p].[Title], [b].[Url]
```

**Comment by puschie286** [comment]
notice that checking for IOrderedQueryable is not enough. But how would you do this when query part 1 (50-53) and part 2 (55-58) doesnt know about each other ? - both have an IQueryable<Blog> as input and output

**Comment by puschie286** [comment]
Why does select remove the ordering ? @cincuranet 
The only purpose of the select is to get the required type for the IQueryable - if there is a better solution, i would be very pleased if you could give a hint or a direction to how this can be solved.

The main problem seems to be that the IQueryable ( that is typed to provide type-safety ) has to return the same type it started (hence the select).

I would like to understand how this simple sql statement can be constructed with linq. currently it seems that generating the sql directly is the only way to achieve this

**Comment by puschie286** [comment]
To make it more pragmatic: how would you write a linq query that should return 'invoice' with multiple join on different entities and order by an property of each joined entity.

(Example: Get all invoice ordered by name of contractor and client location) 

Even if you chain all data in the resultSelector of join, you would end up with selecting all entites (you only want the 'invoice' entity) or lose all ordering.

**Comment by cincuranet** [comment]
You can (have to) write the query as:
```csharp
context.Set<Blog>()
	.Join(context.Set<Post>(), outer => new { Id = outer.BlogId }, inner => new { Id = inner.BlogId }, (outer, inner) => new { o = outer, i = inner })
	.OrderBy(x => x.i.Title)
	.ThenBy(x => x.o.Url)
	.Select(x => x.o);
```

Dynamically composing it would require to know all the orderings in advance to chain it correctly, you can't do it in random places.

**Comment by IanKemp** [comment]
> The check for `IOrderedQueryable<T>` works because the whole query is actually [`EntityQueryable<T>`](https://github.com/dotnet/efcore/blob/main/src/EFCore/Query/Internal/EntityQueryable%60.cs) which implements `IOrderedQueryable<T>` as well, but all this is an internal implementation detail.

That's not an "internal implementation detail", it's a **bug**. If `EntityQueryable<T>` does not actually permit itself to be ordered then it should not implement `IOrderedQueryable<T>`, simple as that.

**Comment by roji** [comment]
@IanKemp despite the forceful wording, things are a bit more complicated than how you make them out to be.

As a LINQ provider, EF returns its implementation of IQueryable (called EntityQueryable) every time a LINQ operator is composed on top of the query. When that operator happens to be OrderBy, that implementation must be castable to IOrderedQueryable, since that's what the OrderBy operator returns - [see the source code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs#L751).  That's why EntityQueryable implements IOrderedQueryable.

It's true that EF's implementation of CreateQuery could check the return type of the method for the node being created, and return a different type (OrderedEntityQueryable) when that method returns IOrderedQueryable. This would align the runtime type we return with the static typing of the operators being composed. It would also add a runtime reflection check for each LINQ operator composition, for all operators and for all users of EF, for something that, well, nobody actually cares about,

I've opened #37311 to track this in the backlog, but I'm not sure I see ourselves making this change. But let me know if you have more thoughts to contribute on this.

**Comment by IanKemp** [comment]
It seems this issue has actually come up before https://github.com/dotnet/efcore/issues/27330 so this might actually be a dupe.

> It would also add a runtime reflection check for each LINQ operator composition, for all operators and for all users of EF, for **something that, well, nobody actually cares about,**

[This question](https://stackoverflow.com/q/24055303/70345) shows that people have had to care about this issue **for over a decade**. Jon Skeet himself notes in his [answer](https://stackoverflow.com/a/24055536/70345) there (emphasis mine);

> Apparently, the code below doesn't correctly determine whether the query is already ordered; even the unordered version implements `IOrderedQueryable<T>`, apparently. Using `source.Expression.Type` to check for `IOrderedQueryable<T>` apparently works, but I've left this code here as the more logical approach - and as **this "fix" sounds more brittle than I'd like** for an answer with my name on it :)

This is one of the the biggest names in C# saying, in the nicest way possible (because that's how he is), that EF's API makes _no sense_ in this regard. I'm just being a little more direct about it, and I'll be even more direct now: **this issue is not something that people working with EF in 2025 should still have to bang their heads against**.

_edit:_ another example from ye olde `System.Web` https://github.com/microsoft/referencesource/blob/ec9fa9ae770d522a5b5f0607898044b7478574a3/System.Web/UI/WebControls/ModelDataSourceView.cs#L467

>  //Bug 180907: We would like to auto sort on DataKeyName when paging is enabled and result is not already sorted by user to overcome a limitation in EF.

Dong a blame on that file indicates it was committed in 2015 so again, this is a pain point that's existed for a decade _and_ has been encountered by Microsoft employees.

I know that's not likely to change your mind or modify the priority of #37311, but I want you to understand that this is not just me being a PITA or dogmatic for the sake of it.

**Comment by roji** [comment]
@IanKemp good find with #27330.

I agree that the current behavior isn't ideal - that's why I opened #37311. However, as you yourself pointed out, this issue has been open since 2015, and there have been almost zero actual complaints about it. Sure, it has been flagged by Jon Skeet, but otherwise it simply seems to hardly affect anyone. So while I agree it's a wart that should ideally be fixed (hence #37311), it simply seems very low-priority, is all.
